### PR TITLE
[FEATURE] Dynamic form for joined fields

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

This PR add a new UI option allowing to activate a dynamic form for joined fields. 

When this option is activated, joined fields are automatically updated according to the "target field".

optimised

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgsvectorlayerjoinbuffer.cpp [hunk 1]
```diff
@@ -392,6 +394,44 @@ const QgsVectorLayerJoinInfo *QgsVectorLayerJoinBuffer::joinForFieldIndex( int i
...
+}
+
+QgsFeature QgsVectorLayerJoinBuffer::joinedFeatureOf( const QgsVectorLayerJoinInfo &info, const QgsFeature &feature ) const
+{
+  QgsFeature joinedFeature;
+
+  if ( info.joinLayer() )
+  {
+    const QVariant targetValue = feature.attribute( info.targetFieldName() );
+    const QString filter = QString( "\"%1\" = %2" ).arg( info.joinFieldName(), targetValue.toString() );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: Better use the QgsExpression quotedColumnRef/quotedValue methods here. That'll safely handle more edge cases.

File: src/gui/qgsattributeform.cpp [hunk 2]
```diff
@@ -1930,3 +1935,50 @@ void QgsAttributeForm::ContainerInformation::apply( QgsExpressionContext *expres
...
+    {
+      Q_FOREACH ( const QString &field, *subsetFields )
+      {
+        QString prefixedName = info->prefixedFieldName( field );
+        changeAttribute( prefixedName, joinFeature.attribute( field ) );
+      }
+    }
+    else
+    {
+      for ( int i = 0; i < joinFeature.fields().count(); i++ )
```
Comment (nyalldawson) [inline src/gui/qgsattributeform.cpp]: should be able to Q_FOREACH through joinFeature.fields() here

**Review by nyalldawson — COMMENTED** [review]
Looking good - got some unit tests coming I hope?

File: src/gui/qgsattributeform.cpp (hunk 2 above)
Comment (pblottiere) [inline src/gui/qgsattributeform.cpp]: Indeed, thank you!

File: src/core/qgsvectorlayerjoinbuffer.cpp (hunk 1 above)
Comment (pblottiere) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: Done

**Comment by pblottiere** [comment]
> got some unit tests coming I hope?

It's ongoing :)

**Comment by pblottiere** [comment]
@nyalldawson 

I added some tests. Let me know what you think.

**Comment by nyalldawson** [comment]
Code looks good to me! I'll let @3nids or @m-kuhn make the call to merge here as this is their area of maintenance.

File: src/core/qgsvectorlayerjoinbuffer.cpp [hunk 3]
```diff
@@ -392,6 +394,46 @@ const QgsVectorLayerJoinInfo *QgsVectorLayerJoinBuffer::joinForFieldIndex( int i
   return &( mVectorJoins[sourceJoinIndex] );
 }
 
+QList<const QgsVectorLayerJoinInfo *> QgsVectorLayerJoinBuffer::joinsWhereFieldIsId( const QgsField &field ) const
+{
+  QList<const QgsVectorLayerJoinInfo *> infos;
+
+  for ( int i = 0; i < mVectorJoins.count(); i++ )
```
Comment (m-kuhn) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: Can that be iteratorized (instead of slower index based looping)

File: src/app/qgsvectorlayerproperties.cpp [hunk 4]
```diff
@@ -1235,16 +1235,21 @@ void QgsVectorLayerProperties::addJoinToTreeWidget( const QgsVectorLayerJoinInfo
     joinItem->setText( 3, QChar( 0x2714 ) );
   }
 
-  joinItem->setText( 4, join.prefix() );
+  if ( join.isDynamicFormEnabled() )
+  {
+    joinItem->setText( 4, QChar( 0x2714 ) );
```
Comment (m-kuhn) [inline src/app/qgsvectorlayerproperties.cpp]: Note to self, maybe we could have some #define or enum for unicode chars used in the UI

File: src/ui/qgsjoindialogbase.ui [hunk 5]
```diff
@@ -126,6 +126,13 @@
      </property>
     </widget>
    </item>
+   <item row="5" column="0">
+    <widget class="QCheckBox" name="mDynamicFormCheckBox">
+     <property name="text">
+      <string>Dynamic form</string>
```
Comment (m-kuhn) [inline src/ui/qgsjoindialogbase.ui]: Can you add a tooltip? I would have guessed that this would make fields updatable. (my fault but just to say that the name wasn't totally obvious to me)

File: src/app/qgsjoindialog.cpp [hunk 6]
```diff
@@ -42,6 +42,8 @@ QgsJoinDialog::QgsJoinDialog( QgsVectorLayer *layer, QList<QgsMapLayer *> alread
 
   mTargetFieldComboBox->setLayer( mLayer );
 
+  mDynamicFormCheckBox->setToolTip( tr( "This option allows values of the joined fields to be automatically updated when the \"Target Field\" is updated" ) );
```
Comment (m-kuhn) [inline src/app/qgsjoindialog.cpp]: Maybe replace the first "updatet" with reloaded or adjusted (better suggestions welcome) and the second one with "changed" to avoid confusion with the concept of an update in db speak where it's changed in the data provider? cc @DelazJ

**Review by m-kuhn — COMMENTED** [review]
Nice feature. Looking forward 👍

**Comment by pblottiere** [comment]
@m-kuhn

Thank you for your review!

I took into account your comments. Let me know if it suits you.

**Comment by pblottiere** [comment]
Travis build is failing due to some tests on task manager. 

So it doesn't seem to be related with this PR.

File: src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp [hunk 7]
```diff
@@ -49,6 +49,11 @@ QVariant QgsEditorWidgetWrapper::defaultValue() const
   return mDefaultValue;
 }
 
+void QgsEditorWidgetWrapper::resetValue()
+{
+  setValue( defaultValue() );
```
Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp]: Now, if a joined feature does not exist, the default value will be shown?

Imagining a joined feature with a "creation date" with default "now()"... I think in this case if the foreign key field is NULL, the creation_date should also be NULL (rather than today's date), no?

Comment (pblottiere) [inline src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp]: Yes you're right, bad idea...

So, if a joined feature does not exist, I just update the corresponding widgets' values with NULL. Do you agree with that?

Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp]: Yes, I think that's the right thing to do (showing a message like "feature does not exist" or so will be considerably harder to get right I guess)

**Comment by pblottiere** [comment]
> Yes, I think that's the right thing to do (showing a message like "feature does not exist" or so will be considerably harder to get right I guess)

As I didn't want to interfere with constraints messages, I added a tooltip on the widget editor in case of a non existent joined feature.

File: src/gui/qgsattributeform.cpp [hunk 8]
```diff
@@ -1945,6 +1946,7 @@ void QgsAttributeForm::updateJoinedFields( const QgsEditorWidgetWrapper &eww )
   if ( infos.count() == 0 || !currentFormFeature( formFeature ) )
     return;
 
+  const QString errToolTip = QStringLiteral( "Joined feature does not exist" );
```
Comment (m-kuhn) [inline src/gui/qgsattributeform.cpp]: Make this `tr()`-anslatable

File: src/gui/qgsattributeform.cpp [hunk 9]
```diff
@@ -223,14 +223,15 @@ void QgsAttributeForm::setMode( QgsAttributeForm::Mode mode )
...
-void QgsAttributeForm::changeAttribute( const QString &field, const QVariant &value )
+void QgsAttributeForm::changeAttribute( const QString &field, const QVariant &value, const QString &toolTip )
 {
   Q_FOREACH ( QgsWidgetWrapper *ww, mWidgets )
   {
     QgsEditorWidgetWrapper *eww = qobject_cast<QgsEditorWidgetWrapper *>( ww );
     if ( eww && eww->field().name() == field )
     {
       eww->setValue( value );
+      eww->widget()->setToolTip( toolTip );
```
Comment (m-kuhn) [inline src/gui/qgsattributeform.cpp]: How about

`eww->setHint( const QString& )` and add a default implementation that does what you do above.

Then in some widgets like "Text Edit" one could do

```
QgsTextEditWidgetWrapper::setHint( const QString& hintText ) override
{
    if ( mBackupText.isNull() )
        mBackupText = hintText;

    if ( hintText.isNull() )
      widget()->setPlaceholderText( mBackupText );
    else
      widget()->setPlaceholderText( hintText );
}
```

Comment (pblottiere) [inline src/gui/qgsattributeform.cpp]: OK so I added a `setHint` method on editor widgets. 

However I'm not sure to understand the aim of the `mBackupText` as there's also a `mPlaceholderText` attribute used for the line edit widget?

Comment (m-kuhn) [inline src/gui/qgsattributeform.cpp]: Not sure either, just thought it might be used by people using .ui forms or custom python logic

Comment (pblottiere) [inline src/gui/qgsattributeform.cpp]: So do you want me to change something else to merge this PR?

Comment (m-kuhn) [inline src/gui/qgsattributeform.cpp]: Sorry, I totally misinterpreted your last message, forget my last message.

What do you think about replacing the placeholder text (on the widget, not the wrapper) temporarily in this case while setting the content to empty?

Comment (pblottiere) [inline src/gui/qgsattributeform.cpp]: Mmm I'm not convinced by this kind of hidden logic. I think it would be strange to finally get a placeholder text after having explicitely set the content to empty.

**Comment by pblottiere** [comment]
@m-kuhn 

The hint is now considered as a placeholder in case of the text edit wrapper. And indeed I've needed to keep the default placeholder and then restore it while setting the hint to empty :).

Let me know if it's what you had in mind.

Thank you!

**Comment by m-kuhn** [comment]
If you like the behavior, it's fine for me ;)

Just from top of my head: I think there is a reason to set a placeholder text for either an example of text that could be entered in a field or for instructions which text to enter in a field. Both require an editable field which (I think) is never the case for a joined field. On a joined field, the placeholder will just be shown when there is an empty text in the field (and not even for NULL values).

In short, I think there is no user setting to protect here. But this behavior should still only be introduced if it looks good from a user perspective.

Your call, let me know if I should hit the merge button :)

**Comment by pblottiere** [comment]
I don't have a strong opinion here...

@haubourg What do you think from a user point of view?

**Comment by haubourg** [comment]
@pblottiere compiling right now, I need to use it to avoid saying dumb things.

**Comment by haubourg** [comment]
Well, testing the PR (commit 0959c11d035f48bea386512f556cec8c6b35147f):
- I don't see any placeholder, only the tooltip when joined layer has no matching row. My settings display "NULL", is that conflicting with the placeHolder? 

- I agree with Matthias that a placeholder should be used to indicate what kind of text input / format is expected. 

I think the tooltip behavior is fine. I would probably reword it from:
"Joined feature does not exist" 
to
" There is no matching feature for [source field name/alias] + [source field value]  in " + [joinedlayertitle]"

If the variables are not accessible easily let's keep it simple with:
"There is no matching feature in the joined layer"

Anyway, I've been playing hard with it and it's working well and pretty stable an dynamic. Kudo's to Paul.

**Comment by m-kuhn** [comment]
I guess 99% of the time that the tooltip/placeholder/whatever text is shown it's simply because the join field is set to `NULL`. I'd therefore go for a simple *"No feature joined"* by default and only go for "there is no matching [...]" if the join field is not null but still no joined feature is found (and can even be regarded as a warning).

**Comment by haubourg** [comment]
@m-kuhn from what I see, if the joined layer has a NULL value, no tooltip is displayed. It displays the tooltip only if there is no matching row in joined layer. 
@pblottiere  Just changing the message (tooltip or placeholder) is enough for me to merge .

**Comment by m-kuhn** [comment]
I was mainly referring to when the foreign key field is set to NULL.

Let's say the joined layer is countries:

"NULL not found on layer countries": a bit strange as message (Even if there was a country with a name NULL it shouldn't be joined if we follow db semantics)

"No country selected" or similar sounds more appropriate in this case.

**Comment by pblottiere** [comment]
@m-kuhn I updated the tooltip in a simple "No feature joined" as you proposed in a previous comment.

And as @haubourg was agree on the current behaviour, I think this PR may be merged?

**Comment by m-kuhn** [comment]
Nice improvements @pblottiere ! This comes in very handy.
